### PR TITLE
Formulate microbit api used by our editor

### PR DESCRIPTION
**Pull request by kory33**
I would like some comments on the design of the API, especially on the following points:

 - [x] specification of `execute` ; is it running on a REPL?
 - [x] specification of `interrupt` ; I added this because I saw it on a mock-up @ZZYSonny created. What does it do exactly?
 - [x] on definition of `MicrobitOutput` ; is there more we can get from the micro:bit?

### CODE REVIEW

**Comment by ZZYSonny** [comment]
1. `execute` is removed. Now we can concentrate on flashing only and this makes tracking state of microbit easier.
2. `interrupt` is like ctrl-C.
    For example, this runs forever until a ctrl-C occurs
```
while True:
    print(1)
```
3. 
- Parsed error messages. Which line, what error. 
- Possibly query state of microbit. Is it flashing? Is it running flashing code? Is it finished?

**Comment by ZZYSonny** [comment]
The api should also manage connection of microbit. Specifically connecting port must be triggered after a user gesture.

**Comment by kory33** [comment]
@ZZYSonny 

I've added interfaces to manage connections.
From what I've seen on MakeCode editor, the browser natively shows a dialog to select the device, hence the signature `connect: () => Promise<InteractWithConnectedMicrobit>`. Are we using the same feature on this editor?

**Comment by ZZYSonny** [comment]
> From what I've seen on MakeCode editor, the browser natively shows a dialog to select the device, hence the signature `connect: () => Promise<InteractWithConnectedMicrobit>`. Are we using the same feature on this editor?

Yes. So `connect` includes the process of user choosing a serial port.

```ts
  /**
   * A promise that completes when the micro:bit connection is no longer active.
   */
  awaitDisconnection: Promise<void>
```
Does this trigger a disconnection?

**Comment by kory33** [comment]
> Does this trigger a disconnection?

No, it's just a promise that's waiting for the micro:bit to disconnect, so if we `await awaitDisconnection` somewhere in our UI we can switch back to a state saying "micro:bit not connected".

Maybe the property name is unnatural when used with `await`, so I'm changing it to `disconnection` and updating doc.

**Comment by ZZYSonny** [comment]
```ts
/**
 * A content that is output to the standard output of micro:bit
 */
export interface NormalOutput {
    kind: 'NormalOutput'
    outputLine: string
}
```
outputLine isn't great here. Suppose you want to read int. Then the first print never gets to the stream.
```python
print("Please give me a int") //not exact function.
a = readInt() //not exact function
```

In my current implementation, I am buffering serial output, until I see lines indicating execution is done. So something more natural could be update on whole program output. For any incremental way, I don't have a good way to know in advance, if it is part of the signal.

I will try make a commit on this.

**Comment by kory33** [comment]
@ZZYSonny I think the API is now able to express state of the flashing with 6c39739, because
 - the returned stream can tell the progress of flashing, and once it returns `DoneFlashing` object we know flashing is done
 - when the user outputs anything or an error occurs, the `outputStream: Stream<MicrobitOutput>`provided by `DoneFlashing` can notify the contents
 - when the process terminates, `outputStream` can terminate

**Comment by kory33** [comment]
@ZZYSonny
I think it's perfectly fine if we cannot separate lines.
If you are buffering the output, I think we can change `outputLine: string` to `outputChunk: string`,
so that every time you get some chunked output from the serial you can put it on the stream.

Also, are we also supporting the inputs to the process now or can that be postponed until later?

**Comment by ZZYSonny** [comment]
It is probably more natural to reboot after flashing. Also flashing is generally quick, so possibly we only need two state DoneFlashing and DoneRebooting?

We should probably draw a diagram on the state change.

**Comment by ZZYSonny** [comment]
> If you are buffering the output, I think we can change `outputLine: string` to `outputChunk: string`,

outputChunk is a better name. 

> Also, are we also supporting the inputs to the process now or can that be postponed until later?

I am planning to support this but I don't find a micropython library function for reading yet.

**Comment by kory33** [comment]
@ZZYSonny 

> It is probably more natural to reboot after flashing. Also flashing is generally quick, so possibly we only need two state DoneFlashing and DoneRebooting?

This is reasonable, I'm making the change!

**Comment by ZZYSonny** [comment]
> @ZZYSonny
> 
> > It is probably more natural to reboot after flashing. Also flashing is generally quick, so possibly we only need two state DoneFlashing and DoneRebooting?
> 
> This is reasonable, I'm making the change!

Actually it is better to make it one state DoneFlashingAndRebooting. There isn't a good signal between flashing and rebooting. 
And reboot action should also produce a outputStream.

**Comment by kory33** [comment]
> Actually it is better to make it one state DoneFlashingAndRebooting. There isn't a good signal between flashing and rebooting.
> And reboot action should also produce a outputStream.

If that's the case I'd just revert 6c39739 since we aren't needing anything more than "flash complete" state,
which is indicated by completion of the promise that `flash` returns.

File: src/api/microbit-api.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+/**
+ * Data that we expect to receive from micro:bit as a result of execututing the flashed code.
+ */
+export type MicrobitOutput = NormalOutput | ErrorMessage
+
+/**
+ * A content that is output to the standard output of micro:bit
+ * outputChunk is the complete output from the flashed program
```
Comment (kory33) [inline src/api/microbit-api.ts]: @ZZYSonny I would expect only a piece of output by a name of "outputChunk". I think we should either
- keep the name `outputChunk`, and only write a piece of new output to the stream
- change the name to `outputSoFar` and put all the output to this string

Comment (ZZYSonny) [reply src/api/microbit-api.ts]: I will be fine with either.

**Comment by ZZYSonny** [comment]
Added `ConnectionFailure` for the case when it fails to get a serial port etc.

```
connect: () => Promise<InteractWithConnectedMicrobit|ConnectionFailure>
```
It this supposed to return a MicrobitConnection?

File: src/api/microbit-api.ts (hunk 1 above)
Comment (kory33) [reply src/api/microbit-api.ts]: right, then let's pick the former and let the UI component append outputs

**Comment by kory33** [comment]
> It this supposed to return a MicrobitConnection?

oops, that's right, it should be `connect: () => Promise<MicrobitConnection|ConnectionFailure>`

**Comment by ZZYSonny** [comment]
Remember that we need to support multiple editors in one page, sharing one MicrobitConnection object. How are you planning to implement this in the component part. A static or global member?

**Comment by kory33** [comment]
> Remember that we need to support multiple editors in one page, sharing one MicrobitConnection object. How are you planning to implement this in the component part. A static or global member?

The connection object can be held at the top-level component, which is above "flash" buttons etc. and list of editors.

**Comment by kory33** [comment]
LGTM at this point.

**Comment by ZZYSonny** [comment]
Final questions
Do we need to add a interact function for sending keys in advance?

Just to confirm the outer editor will maintain the state, so no state checking will be done inside this api. e.g. Sending key when flashing is not possible.

**Comment by kory33** [comment]
> Do we need to add a interact function for sending keys in advance?

Could you explain what do keys mean in this context?

**Comment by ZZYSonny** [comment]
> > Do we need to add a interact function for sending keys in advance?
> 
> Could you explain what do keys mean in this context?

That is "inputs to the process". It can be done by sending individual characters to the terminal. 
```
print("Input integer")
a=readInt  //suppose function exist
```

**Comment by kory33** [comment]
> > > Do we need to add a interact function for sending keys in advance?
> > 
> > 
> > Could you explain what do keys mean in this context?
> 
> That is "inputs to the process". It can be done by sending individual characters to the terminal.
> 
> ```
> print("Input integer")
> a=readInt  //suppose function exist
> ```

I see.

Until we are sure how we can input to the process we could postpone adding in this function.
Otherwise, I'd like to know if we somehow sending the input before flashing or after reboot (so while the process is running).

If it's the latter,

> Just to confirm the outer editor will maintain the state, so no state checking will be done inside this api. e.g. Sending key when flashing is not possible.

we may enforce this restriction on API by returning

```Typescript
export interface ActiveMicrobitProcess {
  sendInput(str: string): Promise<void>
  outputStream: Stream<MicrobitOutput>
}
```
when the promise on `flash` completes, instead of directly returning `outputStream`.

**Comment by ZZYSonny** [comment]
Let us ask industrial sponsor about "text input to microbit" in the meeting today. Then we can finalize the api on flash and reboot. (btw I think `sendInput` on a single character is more appropriate, if you want to have input and output in a same textarea, like the behavior of a terminal app.)
I am happy with all other part.

**Comment by kory33** [comment]
> Let us ask industrial sponsor about "text input to microbit" in the meeting today. Then we can finalize the api on flash and reboot. (btw I think `sendInput` on a single character is more appropriate, if you want to have input and output in a same textarea, like the behavior of a terminal app.)
> I am happy with all other part.

Sure, I think that can be added in the later version of the api, so can I merge this PR for now?

**Comment by ZZYSonny** [comment]
> > Let us ask industrial sponsor about "text input to microbit" in the meeting today. Then we can finalize the api on flash and reboot. (btw I think `sendInput` on a single character is more appropriate, if you want to have input and output in a same textarea, like the behavior of a terminal app.)
> > I am happy with all other part.
> 
> Sure, I think that can be added in the later version of the api, so can I merge this PR for now?

Sure

**Review by ZZYSonny — APPROVED** [review]

**Pull request merged**